README.md
[![Build Status](https://travis-ci.org/SPBTV/rfc5646.svg)](https://travis-ci.org/SPBTV/rfc5646)
[![Gem Version](https://badge.fury.io/rb/rfc5646.svg)](https://badge.fury.io/rb/rfc5646)

# Rfc5646

Parse [RFC 5646](http://tools.ietf.org/html/rfc5646) locales. 
This code extracted from [shuttle](https://github.com/square/shuttle) codebase.

Add this line to your application's Gemfile:

```ruby
gem 'rfc5646'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rfc5646

## Usage

```ruby
locale = Rfc5646::Locale.from_rfc5646(params[:locale])
```

The ISO 639 code for the base language (e.g., "de" for German).

```ruby
locale.iso639
```

The RFC 5646 code for the orthography (e.g., "Arab" for Arabic script).

```ruby
locale.script
```

The ISO 3166 country code for the regional dialect (e.g., "BZ" for Belize). Some special 
values are also supported (e.g., "013" for Central America).

```ruby
locale.region
```

The variant or nested subvariant of this locale. The full path to a subvariant is listed as 
a top-level Array; an example is `["sl", "rozaj", "1994"]`, indicating the 1994 standardization of the 
Resian orthography of the Rozaj dialect of Slovenian (in case we should ever want to localize one of 
our projects thusly). Variants can be regional or temporal dialects, or orthographies, or both, 
and are very specific.

```ruby
locale.variants
```

The dialect (not associated with a specific region or period in time) specifier. For example, 
"yue" indicates Yue Chinese (Cantonese).

```ruby
locale.extended_language
```

The user-defined extensions applied to this locale. The meaning of these is not specified in 
the spec, and left up to private use, and is ignored by this class, but stored for completeness.

```ruby
locale.extensions
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/SPBTV/rfc5646. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org/) code of conduct.


## License

The gem is available as open source under the terms of the [Apache 2.0](http://www.apache.org/licenses/).



CHANGELOG.md
0.1.3

  * Add legal notice on top of locale.rb as it required by Apache 2.0 license

0.1.2

  * Add locales.en.yml to I18n.load_path instead of loading it

0.1.1

  * **Fix** localization loading
  * Allow to pass nil into Locale.from_rfc5646 method


spec/spec_helper.rb
require 'rfc5646'
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # The settings below are suggested to provide a good initial experience
  # with RSpec, but feel free to customize to your heart's content.
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = 'spec/examples.txt'

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
end


spec/rfc5646/locale_spec.rb
RSpec.describe Rfc5646::Locale do
  Locale = Rfc5646::Locale
  I18n.config.available_locales = [:en]

  describe '.from_rfc5646' do
    it 'should recognize a simple locale subtag' do
      locale = Locale.from_rfc5646('de')
      expect(locale.iso639).to eql('de')
      expect(locale.script).to be_nil
      expect(locale.region).to be_nil
      expect(locale.variants).to eql([])
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale + script subtag' do
      locale = Locale.from_rfc5646('zh-Hant')
      expect(locale.iso639).to eql('zh')
      expect(locale.script).to eql('Hant')
      expect(locale.region).to be_nil
      expect(locale.variants).to eql([])
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a an extended locale subtag' do
      locale = Locale.from_rfc5646('zh-cmn-Hans-CN')
      expect(locale.iso639).to eql('zh')
      expect(locale.script).to eql('Hans')
      expect(locale.region).to eql('CN')
      expect(locale.variants).to eql([])
      expect(locale.extended_language).to eql('cmn')
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale-script-region subtag' do
      locale = Locale.from_rfc5646('zh-Hans-CN')
      expect(locale.iso639).to eql('zh')
      expect(locale.script).to eql('Hans')
      expect(locale.variants).to eql([])
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale-variant subtag' do
      locale = Locale.from_rfc5646('sl-rozaj')
      expect(locale.iso639).to eql('sl')
      expect(locale.script).to be_nil
      expect(locale.region).to be_nil
      expect(locale.variants).to eql(%w(rozaj))
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale-region-variant subtag' do
      locale = Locale.from_rfc5646('de-CH-1901')
      expect(locale.iso639).to eql('de')
      expect(locale.script).to be_nil
      expect(locale.region).to eql('CH')
      expect(locale.variants).to eql(%w(1901))
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale-script-region-variant subtag' do
      locale = Locale.from_rfc5646('hy-Latn-IT-arevela')
      expect(locale.iso639).to eql('hy')
      expect(locale.script).to eql('Latn')
      expect(locale.region).to eql('IT')
      expect(locale.variants).to eql(%w(arevela))
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should recognize a locale-region subtag' do
      locale = Locale.from_rfc5646('de-DE')
      expect(locale.iso639).to eql('de')
      expect(locale.script).to be_nil
      expect(locale.region).to eql('DE')
      expect(locale.variants).to eql([])
      expect(locale.extended_language).to be_nil
      expect(locale.extensions).to eql([])
    end

    it 'should return `nil` when `nil` given' do
      locale = Locale.from_rfc5646(nil)
      expect(locale).to be_nil
    end

    it 'should return `nil` when empty string given' do
      locale = Locale.from_rfc5646('')
      expect(locale).to be_nil
    end
  end

  describe '#rfc5646' do
    it 'should properly normalize locale component values' do
      expect(Locale.from_rfc5646('hy-Latn-IT-arevela-x-foobar').rfc5646).to eql('hy-Latn-IT-arevela')
    end
  end

  describe '#name' do
    it 'should properly name all examples from the spec' do
      {
        'de'                    => 'German',
        'fr'                    => 'French',
        'ja'                    => 'Japanese',
        'zh-Hant'               => 'Chinese (Han (Traditional variant) orthography)',
        'zh-Hans'               => 'Chinese (Han (Simplified variant) orthography)',
        'sr-Cyrl'               => 'Serbian (Cyrillic orthography)',
        'sr-Latn'               => 'Serbian (Latin orthography)',
        'zh-cmn-Hans-CN'        => 'Mandarin Chinese (as spoken in China, Han (Simplified variant) orthography)',
        'cmn-Hans-CN'           => 'Mandarin Chinese (as spoken in China, Han (Simplified variant) orthography)',
        'zh-yue-HK'             => 'Yue Chinese (as spoken in Hong Kong)',
        'yue-HK'                => 'Yue Chinese (as spoken in Hong Kong)',
        'zh-Hans-CN'            => 'Chinese (as spoken in China, Han (Simplified variant) orthography)',
        'sr-Latn-RS'            => 'Serbian (as spoken in Serbia, Latin orthography)',
        'sl-rozaj'              => 'Slovenian (Resian)',
        'sl-rozaj-biske'        => 'Slovenian (The San Giorgio dialect of Resian)',
        'sl-nedis'              => 'Slovenian (Natisone dialect)',
        'de-CH-1901'            => 'German (Traditional German orthography as spoken in Switzerland)',
        'sl-IT-nedis'           => 'Slovenian (Natisone dialect as spoken in Italy)',
        'hy-Latn-IT-arevela'    => 'Armenian (Eastern Armenian as spoken in Italy, Latin orthography)',
        'de-DE'                 => 'German (as spoken in Germany)',
        'en-US'                 => 'English (as spoken in United States)',
        'es-419'                => 'Spanish (as spoken in Latin America and the Caribbean)',
        'de-CH-x-phonebk'       => 'German (as spoken in Switzerland)',
        'az-Arab-x-AZE-derbend' => 'Azerbaijani (Arabic orthography)'
      }.each do |subtag, name|
        expect(Locale.from_rfc5646(subtag).name).to eql(name)
      end
    end
  end

  describe '#child_of?' do
    it 'should claim that en-US is a child of en' do
      expect(Locale.from_rfc5646('en-US')).to be_child_of(Locale.from_rfc5646('en'))
    end

    it 'should not claim that ja is a child of en' do
      expect(Locale.from_rfc5646('ja')).not_to be_child_of(Locale.from_rfc5646('en'))
    end

    it 'should claim that zh-cmn-Hans-CN is a child of zh-CN' do
      expect(Locale.from_rfc5646('zh-cmn-Hans-CN')).to be_child_of(Locale.from_rfc5646('zh-CN'))
    end

    it 'should not claim that zh-cmn-Hans-CN is a child of zh-yue-Hans' do
      expect(Locale.from_rfc5646('zh-cmn-Hans-CN')).not_to be_child_of(Locale.from_rfc5646('zh-yue-Hans'))
    end

    it 'should claim that sl-rozaj-biske is a child of sl-rozaj' do
      expect(Locale.from_rfc5646('sl-rozaj-biske')).to be_child_of(Locale.from_rfc5646('sl-rozaj'))
    end

    it 'should not claim that sl-rozaj-biske is a child of sl-nedis' do
      expect(Locale.from_rfc5646('en-rozaj-biske')).not_to be_child_of(Locale.from_rfc5646('sl-nedis'))
    end
  end
end


lib/rfc5646/locale.rb
# Copyright 2014 Square Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

require 'active_support/core_ext/object/try'
require 'active_support/core_ext/object/blank'
require 'i18n'

module Rfc5646
  # Represents a locale that can be localized to. Locales are identified by
  # their RFC 5646 code, which can be as simple as just a language (e.g., "en" for
  # English), or arbitrary complex (e.g., "zh-cmn-Hans-CN" for Mandarin Chinese
  # as spoken in China, simplified Han orthography). The entire RFC 5646 spec is
  # supported by this class.
  class Locale
    # @private
    RFC5646_EXTLANG = /(?<extlang>[a-zA-Z]{3})(-[a-zA-Z]{3}){0,2}/
    # @private
    RFC5646_ISO639 = /(?<iso639>[a-zA-Z]{2,3})(-#{RFC5646_EXTLANG.source})?/
    # @private
    RFC5646_RESERVED = /(?<reserved>[a-zA-Z]{4})/
    # @private
    RFC5646_SUBTAG = /(?<subtag>[a-zA-Z]{5,8})/
    # @private
    RFC5646_REGION = /(?<region>([a-zA-Z]{2}|\d{3}))/
    # @private
    RFC5646_VARIANT = /([a-zA-Z0-9]{5,8}|\d[a-zA-Z0-9]{3})/
    # @private
    RFC5646_SCRIPT = /(?<script>[a-zA-Z]{4})/
    # @private
    RFC5646_EXTENSION = /([0-9A-WY-Za-wy-z](-[a-zA-Z0-9]{2,8}){1,})/
    # @private
    RFC5646_LANGUAGE = /(#{RFC5646_ISO639.source}|#{RFC5646_RESERVED.source}|#{RFC5646_SUBTAG.source})/
    # @private
    RFC5646_PRIVATE = /x(?<privates>(-[a-zA-Z0-9]{1,8}){1,})/
    # @private
    RFC5646_FORMAT = /\A#{RFC5646_LANGUAGE.source}(-#{RFC5646_SCRIPT.source})?(-#{RFC5646_REGION.source})?(?<variants>(-#{RFC5646_VARIANT.source})*)(?<extensions>(-#{RFC5646_EXTENSION.source})*)(-#{RFC5646_PRIVATE.source})?\z/ # rubocop:disable Metrics/LineLength

    # @return [String] The ISO 639 code for the base language (e.g., "de" for
    #   German).
    attr_reader :iso639
    # @return [String] The RFC 5646 code for the orthography (e.g., "Arab" for
    #   Arabic script).
    attr_reader :script
    # @return [String] The ISO 3166 country code for the regional dialect (e.g.,
    #   "BZ" for Belize). Some special values are also supported (e.g., "013" for
    #   Central America); see the spec for details.
    attr_reader :region
    # @return [Array<String>] The variant or nested subvariant of this locale.
    #   The full path to a subvariant is listed as a top-level Array; an example
    #   is `["sl", "rozaj", "1994"]`, indicating the 1994 standardization of the
    #   Resian orthography of the Rozaj dialect of Slovenian (in case we should
    #   ever want to localize one of our projects thusly). Variants can be
    #   regional or temporal dialects, or orthographies, or both, and are very
    #   specific.
    attr_reader :variants
    # @return [String] The dialect (not associated with a specific region or
    #   period in time) specifier. For example, "yue" indicates Yue Chinese
    #   (Cantonese).
    attr_reader :extended_language
    # @return [Array<String>] The user-defined extensions applied to this
    #   locale. The meaning of these is not specified in the spec, and left up
    #   to private use, and is ignored by this class, but stored for completeness.
    attr_reader :extensions

    class << self
      # Generates a new instance from an RFC 5646 code.
      #
      # @param [String, nil] ident The RFC 5646 code for the locale.
      # @return [Locale] The instance representing that locale.
      def from_rfc5646(ident)
        ident = String(ident).tr('_', '-')
        return nil unless (matches = RFC5646_FORMAT.match(ident))
        attrs = RFC5646_FORMAT.named_captures.each_with_object({}) do |(name, offsets), hsh|
          hsh[name] = offsets.map { |offset| matches[offset] }.compact
        end

        iso639 = attrs['iso639'].first
        script = attrs['script'].first
        return nil unless iso639
        region = attrs['region'].first
        if (variants = attrs['variants'].first)
          variants = variants.split('-')
          variants.shift
        end
        if (extensions = attrs['extensions'].first)
          extensions = extensions.split('-')
          extensions.shift
        end
        extlang = attrs['extlang'].first

        Locale.new iso639, script, extlang, region, variants || [], extensions || []
      end

      # Returns an array of Locales representing all possible completions given a
      # prefix portion of an RFC 5646 code. The resolution of the resultant array is
      # determined by the resolution if the input prefix. Some examples:
      #
      # * If just the letter "e" is entered, Locales whose ISO 639 codes begin
      #   with the letter "e" will be returned (English, Spanish, etc.). These
      #   Locale instances will have no other fields specified.
      # * If "en-U" is specified, Locale instances representing "en-US" and
      #   "en-UA", among others, will be returned, as well as "en-Ugar" (for all the
      #   sense it makes). "en-US-Ugar" would not be returned, as it is of a higher
      #   resolution than the input.
      #
      # @param [String] prefix A portion of an RFC 5646 code.
      # @param [Fixnum] max A maximum number of completions to return.
      # @return [Array<Locale>] Candidate completions as Locale instances.

      def from_rfc5646_prefix(prefix, max = nil)
        if prefix.include?('-')
          prefix_path = prefix.split('-')
          prefix = prefix_path.pop
          parent_prefix = prefix_path.join('-')
          parent = from_rfc5646(parent_prefix)
          return [] unless parent

          search_paths = if parent.variants.present?
                           # TODO: subvariants
                           []
                         elsif parent.region # only possible completions are variants
                           []
                           # TODO: variants
                         elsif parent.script # can be followed with variant or region
                           %w(locale.region)
                         else # can be followed with script, region, or variant
                           %w(locale.region locale.script)
                         end

          keys = search_paths.map do |path|
            I18n.t(path).select { |k, v| Locale.matches_prefix? prefix, k, v }.keys
          end.flatten
          keys.delete '_END_'
          keys = keys[0, max] if max

          return keys.map { |key| from_rfc5646 "#{parent_prefix}-#{key}" }
        else
          keys = I18n.t('locale.name').select { |k, v| Locale.matches_prefix? prefix, k, v }.keys
          keys = keys[0, max] if max
          return keys.map { |key| from_rfc5646 key }
        end
      end

      # @private
      def matches_prefix?(prefix, key, value)
        return false unless value.is_a?(String)
        return true if key.to_s.downcase.starts_with?(prefix.downcase)
        return true if value.split(/\w+/).any? { |word| word.downcase.starts_with? prefix.downcase }
        false
      end

      private

      def fallbacks
        @fallbacks ||= YAML.load_file(Rails.root.join('data', 'fallbacks.yml'))
      end
    end

    # @private
    def initialize(iso639, script = nil, extlang = nil, region = nil, variants = [], extensions = [])
      @iso639 = iso639.try!(:downcase)
      @region = region.try!(:upcase)
      @variants = variants.map(&:downcase)
      @extended_language = extlang.try!(:downcase)
      @extensions = extensions
      @script = script
    end

    # @return [String] The full RFC 5646 code for this locale.

    def rfc5646
      [iso639, script, extended_language, region, *variants].compact.join('-')
    end

    alias_method :to_param, :rfc5646

    # Returns a human-readable localized name of the locale.
    #
    # @param [String] locale The locale to use (default locale is used by
    #   default).
    # @return [String] The localized name of the locale.

    def name(locale = nil)
      I18n.with_locale(locale || I18n.locale) do
        i18n_language = if extended_language
                          I18n.t "locale.extended.#{iso639}.#{extended_language}"
                        else
                          I18n.t "locale.name.#{iso639}"
                        end

        i18n_dialect = if variants.present?
                         I18n.t "locale.variant.#{iso639}.#{variants.join '.'}._END_"
                       end

        i18n_script = script ? I18n.t("locale.script.#{script}") : nil
        i18n_region = region ? I18n.t("locale.region.#{region}") : nil

        if i18n_region && i18n_dialect && i18n_script
          I18n.t('locale.format.scripted_regional_dialectical',
                 script: i18n_script,
                 dialect: i18n_dialect,
                 region: i18n_region,
                 language: i18n_language
                )
        elsif i18n_region && i18n_dialect
          I18n.t('locale.format.regional_dialectical',
                 dialect: i18n_dialect,
                 region: i18n_region,
                 language: i18n_language
                )
        elsif i18n_region && i18n_script
          I18n.t('locale.format.scripted_regional',
                 script: i18n_script,
                 region: i18n_region,
                 language: i18n_language
                )
        elsif i18n_dialect && i18n_script
          I18n.t('locale.format.scripted_dialectical',
                 script: i18n_script,
                 dialect: i18n_dialect,
                 language: i18n_language
                )
        elsif i18n_script
          I18n.t('locale.format.scripted',
                 script: i18n_script,
                 language: i18n_language
                )
        elsif i18n_dialect
          I18n.t('locale.format.dialectical',
                 dialect: i18n_dialect,
                 language: i18n_language
                )
        elsif i18n_region
          I18n.t('locale.format.regional',
                 region: i18n_region,
                 language: i18n_language
                )
        else
          i18n_language
        end
      end
    end

    # Tests for equality between two locales. Their full RFC 5646 codes must be
    # equal.
    #
    # @param [Locale] other Another Locale.
    # @return [true, false] Whether it is the same Locale as the receiver.
    # @raise [ArgumentError] If `other` is not a Locale.

    def ==(other)
      case other
      when Locale
        rfc5646 == other.rfc5646
      else
        false
      end
    end

    alias_method :eql?, :==
    alias_method :equal?, :==
    alias_method :===, :==

    # @private
    def hash
      rfc5646.hash
    end

    # Returns the fallback order for this Locale. For example, fr-CA might
    # fall back to fr, which then falls back to en. The fallback order is
    # described in the `fallbacks.yml` file.
    #
    # @return [Array<Locale>] The fallback order of this locale, from most
    #   specific to most general. Note that this array includes the receiver.

    def fallbacks
      fallbacks = Array.wrap(self.class.fallbacks[rfc5646])
                  .map { |l| self.class.from_rfc5646 l }
      fallbacks.unshift self
      fallbacks
    end

    # Returns whether this Languge is a subset of the given locale. "en-US" is a
    # child of "en".
    #
    # @param [Locale] parent Another locale.
    # @return [true, false] Whether this locale is a child of `parent`.

    def child_of?(parent)
      return false if iso639 != parent.iso639
      return false if parent.specificity > specificity
      parent.specified_parts.all? { |part| specified_parts.include?(part) }
    end

    # @return [true, false] Whether this locale is a pseudo-locale.
    def pseudo?
      variants.include? 'pseudo'
    end

    # @private
    def specificity
      specificity = 1
      specificity += 1 if script
      specificity += 1 if region
      specificity += variants.size
      specificity += 1 if extended_language
      specificity + extensions.size
    end

    # @private
    def specified_parts
      # relies on the fact that the namespace for each element of the code is
      # *globally* unique, not just unique to the code element
      (variants + extensions + [script, region, extended_language]).compact
    end

    # @private
    def as_json(_options = nil)
      {
        rfc5646: rfc5646,
        components: {
          iso639: iso639,
          script: script,
          extended_language: extended_language,
          region: region,
          variants: variants,
          exensions: extensions
        },
        name: name
      }
    end

    # @private
    def inspect
      "#<#{self.class} #{rfc5646}>"
    end
  end
end
